code/xcvm/evm/src/Gateway.sol

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.14;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+import "./Interpreter.sol";
+import "./interfaces/IGateway.sol";
+
+contract Gateway is Ownable, IGateway {
+    mapping(uint256 => mapping(bytes => address)) public userInterpreter;
+
+    mapping(address => Bridge) public bridgesInfo;
+    // TODO ? do we have only one bridgge per network and security
+    mapping(uint256 => mapping(BridgeSecurity => address)) public bridges;
+    mapping(uint256 => address) public assets;
+
+    event InstanceCreated(uint256 networkId, bytes user, address instance);
+
+    event Spawn(
+        bytes account,
+        uint256 networkId,
+        BridgeSecurity security,
+        uint256 salt,
+        bytes spawnedProgram,
+        address[] assetAddresses,
+        uint256[] amounts
+    );
+
+    modifier onlyBridge() {
+        require(bridgesInfo[msg.sender].security != BridgeSecurity(0));
+        _;
+    }
+
+    constructor() {
+        // enable trustless bridge;
+    }
+
+    function getAsset(uint256 assetId) external returns (address) {
+        return assets[assetId];
+    }
+
+    function getBridge(uint256 networkId, BridgeSecurity security) external returns (address) {
+        return bridges[networkId][security];
+    }
+
+    function registerAsset(address assetAddress, uint128 assetId) external onlyOwner {
+        require(assetAddress != address(0), "Gateway: invalid address");
+        assets[assetId] = assetAddress;
+    }
+
+    function unregisterAsset(uint128 assetId) external onlyOwner {
+        delete assets[assetId];
+    }
+
+    function registerBridge(
+        address bridgeAddress,
+        BridgeSecurity security,
+        uint256 networkId
+    ) external onlyOwner {
```

Comment:
SPEC states that there can be multiple admins of the contract. Router is the default owner and other owners are configurable. I think we should modify `onlyOwner` in this case.

Reply:
It's a good point, I think we should have a set of owners defined by the router but we should ask @KaiserKarel

Reply:
Currently, what I do is router sets the owners while instantiating an interpreter. But I'm not sure if there is something like `instantiate` in solidity but maybe something like `constructor`. And then there is no way to change the owners. Maybe we need that though idk.

Reply:
yeah, this is something that the SPEC currently does not cover at all. We need to prioritize this asap.

Reply:
Explaining the flow here based on the conversations here and slack to be able to resolve this: